dotnet/src/InternalUtilities/connectors/AI/FunctionCalling/FunctionCallsProcessorLoggerExtensions.cs

```diff
@@ -0,0 +1,147 @@
+        {
+            var functionsLog = (configuration.Functions != null && configuration.Functions.Any())
+                ? string.Join(", ", configuration.Functions.Select(f => FunctionName.ToFullyQualifiedName(f.Name, f.PluginName)))
+                : "None (Function calling is disabled)";
+
+            s_logFunctionChoiceBehaviorConfiguration(
+                logger,
+                configuration.Choice.Label,
+                configuration.AutoInvoke,
+                configuration.Options.AllowConcurrentInvocation,
+                configuration.Options.AllowParallelCalls,
+                functionsLog,
+                null);
+        }
+    }
+
+    /// <summary>
+    /// Logs function calls.
+    /// </summary>
+    public static void LogFunctionCalls(this ILogger logger, List<FunctionCallContent> functionCalls)
+    {
+        if (logger.IsEnabled(LogLevel.Debug))
+        {
+            s_logFunctionCalls(
+                logger,
+                string.Join(", ", functionCalls.Select(call => $"{FunctionName.ToFullyQualifiedName(call.FunctionName, call.PluginName)} [Id: {call.Id}]")),
+                null
+            );
+        }
+    }
+
+    /// <summary>
+    /// Logs the <see cref="AutoFunctionInvocationContext"/>.
+    /// </summary>
+    public static void LogAutoFunctionInvocationFilterContext(this ILogger logger, AutoFunctionInvocationContext context)
+    {
+        if (logger.IsEnabled(LogLevel.Debug))
+        {
+            var fqn = FunctionName.ToFullyQualifiedName(context.Function.Name, context.Function.PluginName);
+
+            s_logAutoFunctionInvocationFilterContext(
+                    logger,
+                    fqn,
+                    context.ToolCallId,
+                    context.IsStreaming,
+                    context.FunctionSequenceIndex,
+                    context.RequestSequenceIndex,
+                    context.FunctionCount,
+                    null);
+        }
+    }
+
+    /// <summary>
+    /// Logs the auto function invocation process termination.
+    /// </summary>
+    public static void LogAutoFunctionInvocationProcessTermination(this ILogger logger, AutoFunctionInvocationContext context)
+    {
+        if (logger.IsEnabled(LogLevel.Debug))
+        {
+            var fqn = FunctionName.ToFullyQualifiedName(context.Function.Name, context.Function.PluginName);
+
+            s_logAutoFunctionInvocationFilterTermination(logger, fqn, context.ToolCallId, null);
+        }
+    }
+
+    /// <summary>
+    /// Logs function call request failure.
+    /// </summary>
+    public static void LogFunctionCallRequestFailure(this ILogger logger, FunctionCallContent functionCall, string error)
+    {
+        if (logger.IsEnabled(LogLevel.Debug))
+        {
+            var fqn = FunctionName.ToFullyQualifiedName(functionCall.FunctionName, functionCall.PluginName);
+
+            logger.LogDebug("Function call request failed: Name:{Name}, Id:{Id}", fqn, functionCall.Id);
```

Comment:
Why not log the error for debug mode?  Would you not potentially be trying to debug errors in this case?

Reply:
I thought that the error/exception message might contain sensitive data, such as API keys, emails, addresses, etc., so I log the error with a dedicated log level for this type of data: Trace.